.github/workflows/ci.yaml

```diff
@@ -0,0 +1,73 @@
+name: ci
+
+on:
+  pull_request:
+  push:
+    branches:
+      - master
+
+jobs:
+  unit-testing:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: niden/actions-memcached@v7
+      - name: Restore Go cache
+        uses: actions/cache@v1
+        with:
+          path: ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
+      - name: Setup Go
+        uses: actions/setup-go@v2-beta
+        with:
+          go-version: 1.14.x
+      - name: Run tests
+        run: make test TEST_FLAGS="-race -tags integration -timeout 5m"
+      - name: Check codegen
+        run: make check-generated
+      - name: Build binaries
+        run: make all
+      - name: Prepare cache
+        run: |
+          cp "$(go env GOPATH)/bin/fluxctl" cache/fluxctl
+          docker save -o cache/flux-latest.tar fluxcd/flux:latest
+      - name: Upload cache
+        uses: actions/upload-artifact@v1
+        with:
+          name: cache
+          path: cache
+  e2e-testing:
+    runs-on: ubuntu-latest
+    needs: unit-testing
+    strategy:
+      matrix:
+        test: ['10_* 11_* 12_*', '13_* 14_*', '15_* 16_* 17_*', '20_* 21_* 22_*']
```

Comment:
Would something like

```
        test: ['1/4', '2/4', '3/4', '4/4']
```

be an option (with additional logic added to `run-gh.bash`)? This would make (newly) added tests run without (forgetting) to add them here.

Reply:
I've balanced the tests to take the same amount of time, for example running 15 in shard 2 will take 18m

Reply:
Ack.

Reply:
I think we should refactor the tests and have the balancing logic in bash, but this can come later.